modules/ServerAPI/src/client/Protocol.coffee

```diff
@@ -499,3 +544,580 @@ class ProtocolBaseController extends BaseEntityController
+		codeOrigin: 'ACAS DDict'
+		value: null
+	fieldSettings:
+		fieldType: 'booleanValue'
+		formLabel: ''
+		fieldWrapper: "bv_endpointConditionCheckbox"
+]
+		
+
+class EndpointListController extends AbstractFormController
+	template: _.template($("#EndpointListView").html())
+
+	events:
+		"click .bv_addEndpoint": "handleAddEndpointPressed"
+		"rowRemoved": "handleRowRemoved"
+		"click .bv_endpointRow": "handleEndpointRowPressed"
+		"click .bv_downloadFiles": "downloadFiles"
+	
+	rowNumberKind: "column order"
+	stateType: "metadata"
+	stateKind: "data column order"
+	
+	initialize: (options) =>
+		@model = options.model
+
+		endpointStates = @model.get("lsStates").getStatesByTypeAndKind @stateType, @stateKind
+
+		# Sort the collection by the "column order" values
+		@collection = endpointStates.sort (stateA, stateB) =>
+			rnA = @getRowNumberForState(stateA)
+			rnB = @getRowNumberForState(stateB)
+			return rnA - rnB
+		
+	render: => 
+		$(@el).empty()
+		$(@el).html @template()
+
+		#If the table is read-only, remove the remove and add buttons
+		if @options.readOnly == true
+			#remove column
+			@$(".bv_endpointColumnRemove").remove()
+			#remove add endpoint button
+			@$(".bv_addEndpoint").remove()
+			$(".bv_endpointManagerInstructions").hide()
+		
+		if @options.newProtocol == true
+			@$(".bv_downloadFiles").hide()
+			$(".bv_endpointManagerInstructions").hide()
+
+		#check whether or not to display time and concentration columns in the endpoint table
+		@showTimeAndConcentration()
+		
+		# Create a list to hold the endpoint controllers in, so we can iterate through them later
+		@endpointControllers = []
+		for lsState in @collection
+			@.addOne(lsState)
+			if @options.readOnly == true
+				#hide remove buttons
+				@$(".bv_remove_row").hide()
+
+
+		if @options.readOnly == false || @options.newProtocol == true 	#Don't render associated experiments if protocol is new or readOnly
+			@getExperimentSummaryTable()
+		else #if the table isn't rendered, don't render the download files button either
+			$(".bv_downloadFiles").hide() 
+
+		@
+	
+	showTimeAndConcentration: =>
+		#if time/units or concentration/units are disabled, remove the columns and cells from the endpoint controller whenever table is made or a row added
+		if window.conf.protocol.endpointManager.showTime == false
+			@$(".bv_endpointColumnTime").remove()
+			@$(".bv_endpointColumnTimeUnits").remove()
+			@$(".bv_timeUnitsPickListParent").remove()
+			@$(".bv_timeInputParent").remove()
+
+		if window.conf.protocol.endpointManager.showConcentration == false
+			@$(".bv_endpointColumnConcentration").remove()
+			@$(".bv_endpointColumnConcentrationUnits").remove()
+			@$(".bv_concentrationUnitsPickListParent").remove()
+			@$(".bv_concentrationInputParent").remove()
+
+	getExperimentSummaryTable: =>
+		#hide previously shown warnings/success text
+		$(".bv_downloadSuccess").hide()
+		$(".bv_downloadWarning").hide()
+
+		protocolCode = @model.escape('codeName')		
```

Comment:
I think we need to check that the protocol is saved already, otherwise this will fail, or at least the Ajax request will.

Reply:
That makes sense! I added some checks to check if the protocol is new or not upstream of where this would be hit.